pkg/identityserver/user_registry.go

```diff
@@ -478,39 +481,60 @@ func (is *IdentityServer) updateUser(ctx context.Context, req *ttnpb.UpdateUserR
 				return err
 			}
 		}
-		updatingContactInfo := ttnpb.HasAnyField(req.FieldMask.GetPaths(), "contact_info")
-		var contactInfo []*ttnpb.ContactInfo
+
 		updatingPrimaryEmailAddress := ttnpb.HasAnyField(req.FieldMask.GetPaths(), "primary_email_address")
-		if updatingContactInfo || updatingPrimaryEmailAddress {
-			if updatingContactInfo {
-				contactInfo, err = st.SetContactInfo(ctx, req.User.GetIds(), req.User.ContactInfo)
-				if err != nil {
-					return err
-				}
+		updatingContactInfo := ttnpb.HasAnyField(req.FieldMask.GetPaths(), "contact_info")
+
+		contactInfo, err := st.GetContactInfo(ctx, req.User.GetIds())
+		if err != nil {
+			return err
+		}
+
+		if updatingPrimaryEmailAddress {
+			oldUser, err := st.GetUser(ctx, req.User.GetIds(), []string{"primary_email_address"})
```

Comment:
This assumes that the previous primary email address will be found in the `contact_info`, which is not the case for users which have change their primary email address in the meantime.

An update to the `primary_email_address` without an update to `contact_info` should forcefully re-create the `contact_info` using the new email address in my opinion.

Reply:
It was more under the assumption that it shouldn't be done but took some time to do some changes which I believe it makes more sense.

Reply:
The problem here is that there no guarantee that the contact info email is the old user email basically. If I edited my email on version `v3.27.2`, from `adrian@thethingsindustries.com` to `adrian.mares@thethingsindustries.com`, my email in `oldUser.PrimaryEmailAddress` is `adrian.mares@thethingsindustries.com` while the contact info will contain `adrian@thethingsindustries.com` - the loop bellow won't find it and will just append a new email in the contact info.

You cannot depend on the contact info containing anything related to the 'old' user - you should just always override it as long as the user didn't update it himself.

Reply:
Agree but that is only for the first update after this PR is merged, since https://github.com/TheThingsNetwork/lorawan-stack/blob/b956d4a6af012dab137d4ef0f5942bd6b510fb52/pkg/identityserver/user_registry.go#L537-L547 ensures that if the new email will be added to the contactInfo list if the old email was not, all subsequent operations will have a oldEmail to look into within the contactInfo list.

Reply:
But this is my problem here - there is no point in keeping the old contact info in place. It's an old, unused email which shouldn't be stored - we should just override it.

Reply:
I see, I was trying to avoid overwriting the data in the field until we remove everything in https://github.com/TheThingsNetwork/lorawan-stack/issues/6515 but if its permissible to do so then I agree on just overwriting it